libsql/src/hrana/transaction.rs

```diff
@@ -0,0 +1,64 @@
+use crate::hrana::pipeline::{ExecuteStreamReq, StreamRequest};
+use crate::hrana::proto::{Batch, BatchResult, Stmt, StmtResult};
+use crate::hrana::stream::HttpStream;
+use crate::hrana::{HttpSend, Result};
+use crate::TransactionBehavior;
+
+#[derive(Debug, Clone)]
+pub(crate) struct HttpTransaction<T>
+where
+    T: for<'a> HttpSend<'a>,
+{
+    stream: HttpStream<T>,
+}
+
+impl<T> HttpTransaction<T>
+where
+    T: for<'a> HttpSend<'a>,
+{
+    pub fn stream(&self) -> &HttpStream<T> {
+        &self.stream
+    }
+
+    pub async fn open(stream: HttpStream<T>, tx_behavior: TransactionBehavior) -> Result<Self> {
+        let begin_stmt = match tx_behavior {
+            TransactionBehavior::Deferred => "BEGIN DEFERRED",
+            TransactionBehavior::Immediate => "BEGIN IMMEDIATE",
+            TransactionBehavior::Exclusive => "BEGIN EXCLUSIVE",
+            TransactionBehavior::ReadOnly => "BEGIN READONLY",
+        };
+        stream.execute(Stmt::new(begin_stmt, false)).await?;
+        Ok(HttpTransaction { stream })
```

Comment:
On one side yes. On the other - this may screw happen before relationships as opening the transaction won't actually start a new transaction ie.:

1. Have a database in state S1.
2. Open transaction T1 (current db state: S1).
3. Open transaction T2.
4. Within T2, modify database state S1&rarr;S2.
5. Read something in T1: since this is first time when actual request is being made - T1 will point to S2.